### PR TITLE
Fix dynamic alloc

### PR DESCRIPTION
**Pull request by hellkite500**
Refactors the allocation of Q to ensure the correct size for iterating over the histogram size.

Also fixes an accumulation bug where Q was not previously accounting for the time delayed contributions.

Removes hard coded limits and issues warnings when users supply values that may violate original model assumptions.

A few code cleanup/refactors for simplifying and optmization loops.

A few other small bug fixes and comments added.

Note this PR is dependent on #37.  It is branched from this commit on the PR branch: cfbdec5258dfa7174e50b1aa9efefb9f482b06b0

### CODE REVIEW

**Comment by PhilMiller** [comment]
Given that #37 is a big functional change that's still in draft mode, would it make sense to
* Target this PR against the branch for #37 rather than `master`, and/or
* separate out commits of bug fixes that are applicable directly to `master`, and put them up for review and integration separately?

File: test/Makefile [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # define the C compiler to use
-CC = gcc
+CC = gcc -g
 
 # define any compile-time flags
 # for newer compiler, the right hand side is not needed...
 # ... for instance, remove the flags if using gnu/10.1.0 on Cheyenne.
-CFLAGS =
+CFLAGS = -fsanitize=address
```
Comment (PhilMiller) [inline test/Makefile:7]: Probably best not to commit changes like this back to the shared repository.

**Comment by PhilMiller** [comment]
Incidentally, the `test/main_unit_test_bmi.c` here is a decent illustration of the things that I want to capture in a BMI module usage/conventions checking utility. So, I'm glad I came to look.

**Comment by hellkite500** [comment]
As far as merge/integration is concerned, I'm happy to rebase this again and reorganize the commits as appropriate/needed. I'm not sure a PR against the feature branch of #37 would help significantly, based on discussions with @Ben-Choat I think that work is nearly ready to merge, and that might complicate things. As long as this doesn't go stale, I can rebase on master once the feature is complete.

**Pull request closed without merging**